### PR TITLE
ET-4641 propagate schema changes to es mapping

### PR DESCRIPTION
**Pull request by rustonaut**
- propagate schema updates as property mapping updates to elasticsearch
- run a background job to update the indices

**References:**

- [ET-4641]

[ET-4641]: https://xainag.atlassian.net/browse/ET-4641?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: web-api/src/storage/elastic.rs [hunk 1]
```diff
@@ -356,10 +362,70 @@ impl Client {
         }));
 
         Ok(self
-            .query_with_json::<_, IgnoredResponse>(url, body)
+            .query_with_json::<_, IgnoredResponse>(Method::POST, url, body)
             .await?
             .map(|_| ()))
     }
+
+    pub(super) async fn extend_elasticsearch_mapping(
```
Comment (janpetschexain) [inline web-api/src/storage/elastic.rs:370]: :art: we could skip the name as it is already a function on the es client.
```suggestion
    pub(super) async fn extend_mapping(
```

**Review by janpetschexain — COMMENTED** [review]
we could also use the tooling introduced here instead of the integration test setup changes that i made in #1002. to facilitaty this better and to avoid race conditions, wdyt about making the `wait_for_completion` parameter in the index update configurable such that we can await it in the tests? having the handle would be nicer though. also i guess it shouldn't be much of an issue for the integration tests as the index change should be pretty instantaneous.

File: web-api/src/storage/elastic.rs [hunk 2]
```diff
@@ -356,10 +362,70 @@ impl Client {
...
+            [
+                ("conflicts", Some("proceed")),
+                //FIXME add a way to async query the status of the update
+                //      ES will return a task handle we can use for this.
+                ("wait_for_completion", Some("false")),
+                ("refresh", Some("true")),
+                //TODO make it configurable, currently it's 500_000 documents per second
+                //      (500 batch requests per second * 1000 documents per batch)
+                //     but I have no idea what a good value is
+                ("requests_per_second", Some("500")),
```
[the remark this answers is not in the document]
Comment (rustonaut) [reply web-api/src/storage/elastic.rs:422]: yes I got throw of by this sentences 

> control the **rate at which update by query issues batches** of update operations

emphasis mine, but yes  they indeed define it on a per request basis

:pushpin:  add configuration option the current value is pretty arbitrary

**Comment by rustonaut** [comment]
I add config options for the index update requests per minute parameter and which method it use for the update (background, wait for completion).

**Comment by rustonaut** [comment]
I'm just currently debating of weather we have to make the handling of unusual error wrt. ES/PG  more robust. 

Currently you can have an situation where with certain system error cases you can have no updated index in PG but an updated index in ES which also would make retryies of the operation fail. It's rather unlikely as long as wait for completion isn't used, but still possible in various ways.

We could commit PG before updating ES but then we can have a situation where there is a schema in PG but not ES, which has all the same issues (through in PG we always can differentiate between `keyword[]` and `keyword` and in ES we can not).

**Review by janpetschexain — APPROVED** [review]
👍

**Pull request merged**